common: general improvements

dougsland opened the pull request:
- if timeout happens we try 5 times before sending Timeout error to users.

- improve user experience when timeout occurs

- Add logging to handle message info/warn/error

- Add MIT header

Resolves: https://github.com/containers/ramalama/issues/693

## Summary by Sourcery

Enhancements:
- Improve the user experience when a timeout occurs during download.

dougsland commented:
Tested on macOS, I will give a try on Linux, please wait.

dougsland commented on ramalama/common.py:
```diff
@@ -165,25 +186,63 @@ def download_file(url, dest_path, headers=None, show_progress=True):
+                "- Server is down or unresponsive\n"
+                "- Firewall or proxy blocking the request\n"
+            )
+            sys.exit(1)
```
sys.exit here helps exit without Traceback.

dougsland commented on ramalama/common.py:
```diff
@@ -165,25 +186,63 @@ def download_file(url, dest_path, headers=None, show_progress=True):
+            retries += 1
+            logging.warning(f"⚠️  {e}. Retrying {retries}/{max_retries}...")
+
+        except Exception as e:
```
Trying to handle errors not raised during my tests.

dougsland commented:
On macOS:

```
$ ramalama run deepseek-r1:8b          
 20% |███████████████████████████████                                                                                                                              |  981.34 MB/   4.58 GB   9.36 MB/s    8m 17sWARNING:root:⚠️  TimeoutError: The server took too long to respond. Retrying 1/5...
ERROR:root:🚨 Unexpected error: Network error: [Errno 8] nodename nor servname provided, or not known
ERROR:root:🚨 Unexpected error: Network error: [Errno 8] nodename nor servname provided, or not known
ERROR:root:🚨 Unexpected error: Network error: [Errno 8] nodename nor servname provided, or not known
ERROR:root:🚨 Unexpected error: Network error: [Errno 8] nodename nor servname provided, or not known

❌ Download failed after multiple attempts.
Possible causes:
- Internet connection issue
- Server is down or unresponsive
- Firewall or proxy blocking the request
$
```

dougsland commented:
On Linux output:

```
douglas@fedora:~/linux-test/ramalama$ ramalama run deepseek-r1:8b
  3% |█████                                                                                                                                                                                       |  142.34 MB/   4.58 GB  26.32 MB/s    2m 52s2025-01-31 23:43:52,181 - WARNING - ⚠️  TimeoutError: The server took too long to respond. Retrying 1/5...
2025-01-31 23:43:52,384 - ERROR - 🚨 Unexpected error: Network error: [Errno -3] Temporary failure in name resolution
2025-01-31 23:43:52,785 - ERROR - 🚨 Unexpected error: Network error: [Errno -3] Temporary failure in name resolution
2025-01-31 23:43:53,587 - ERROR - 🚨 Unexpected error: Network error: [Errno -3] Temporary failure in name resolution
2025-01-31 23:43:55,189 - ERROR - 🚨 Unexpected error: Network error: [Errno -3] Temporary failure in name resolution
douglas@fedora:~/linux-test/ramalama$
~                                     
```

dougsland commented:
Holy moly these AI bots on github are getting good, let implement the suggestions.

dougsland commented:
I believe it's ready for review :)

dougsland commented:
CI/CD error not related to the patch.

dougsland commented:
@rhatdan PTAL

ericcurtin commented on ramalama/common.py:
```diff
@@ -165,25 +186,67 @@ def download_file(url, dest_path, headers=None, show_progress=True):
+            logging.info(f"✅ Successfully downloaded {url} to {dest_path}.")
+            return  # Exit function if successful
+
+        except urllib.error.HTTPError as e:
```
I think the fail messages are fine, but I would delete all the successful messages I think.

dougsland replied:
sure

ericcurtin commented on ramalama/common.py:
```diff
@@ -1,17 +1,38 @@
-"""ramalama common module."""
```
I would prefer to avoid putting these licenses headers everywhere, it's a pain, you have to remember to update the year, etc. One license file is sufficient.

dougsland replied:
Oh no,  usually is good to have header license. Linux kernel and others use it very often.

ericcurtin replied:
Linux kernel do a lot of things, I don't think we have to copy them though. PyTorch don't do this and I'm sure we could find many projects that do or don't.

It's not worth the maintenance, we are licensed . We don't do this for any other python file in the project.

ericcurtin replied:
I think Linux kernel community doesn't even recommend this anymore themselves, I think they recommend one-liner SPDX (which I don't think we need either).

It's just unnecessary process and maintenance IMO, unless we are legally advised to do it, I'd rather not.

ericcurtin commented:
We should check if the terminal actually supports emojis before printing them, this is how systemd does it:

https://github.com/systemd/systemd/blob/75ee025c5de5d753dc1d8a28f8780247f5a887ae/src/basic/glyph-util.c#L8

dougsland commented:
> We should check if the terminal actually supports emojis before printing them, this is how systemd does it:
> 
> https://github.com/systemd/systemd/blob/75ee025c5de5d753dc1d8a28f8780247f5a887ae/src/basic/glyph-util.c#L8

I will skip emoji for now I guess to make it less complicated.

ericcurtin commented:
> > We should check if the terminal actually supports emojis before printing them, this is how systemd does it:
> > https://github.com/systemd/systemd/blob/75ee025c5de5d753dc1d8a28f8780247f5a887ae/src/basic/glyph-util.c#L8
> 
> I will skip emoji for now I guess to make it less complicated. :(

It would be something like this FWIW, when we get around to it:

def is_locale_utf8():
    # This is a simplified version. Adjust this function based on your locale check needs.
    return 'UTF-8' in os.getenv('LC_CTYPE', '') or 'UTF-8' in os.getenv('LANG', '')

def emoji_enabled():
    term = os.getenv("TERM")
    if not term or term in ("dumb", "linux"):
        return False

    return is_locale_utf8()

rhatdan commented:
The emojis are fun, but want to make sure they work on Windows, Linux and Mac terminals.

dougsland commented:
> > > We should check if the terminal actually supports emojis before printing them, this is how systemd does it:
> > > https://github.com/systemd/systemd/blob/75ee025c5de5d753dc1d8a28f8780247f5a887ae/src/basic/glyph-util.c#L8
> > 
> > 
> > I will skip emoji for now I guess to make it less complicated. :(
> 
> It would be something like this FWIW, if we get around to it:
> 
> ```
> def is_locale_utf8():
>     return 'UTF-8' in os.getenv('LC_CTYPE', '') or 'UTF-8' in os.getenv('LANG', '')
> 
> def emoji_enabled():
>     term = os.getenv("TERM")
>     if not term or term in ("dumb", "linux"):
>         return False
> 
>     return is_locale_utf8()
> ```

Thanks tonight I will double check and test it.

dougsland commented:
> The emojis are fun, but want to make sure they work on Windows, Linux and Mac terminals.

Already tested on mac and linux, work great. Windows to be tested soon.

ericcurtin commented:
> > The emojis are fun, but want to make sure they work on Windows, Linux and Mac terminals.
> 
> Already tested on mac and linux, work great. Windows to be tested soon.

I think just copy the systemd technique, we won't be able to test every terminal, platform, combination, etc. systemd is deployed at crazy scale, they probably have it done as correctly as it can be within reason

ericcurtin commented on ramalama/console.py:
```diff
@@ -0,0 +1,53 @@
+"""
```
Please remove the license text

dougsland replied:
still disagree here, specially in new files but okay.

ericcurtin commented:
If we create a new file, we must add to install.sh , we have broken installs already because of this with gpu_detector

ericcurtin commented on ramalama/console.py:
```diff
@@ -0,0 +1,53 @@
+    if not is_locale_utf8():
+        return False  # Block emoji if UTF-8 is not in locale
+
+    if sys.platform == "win32": # works 32, 64 bits machines
```
We can remove this block, we don't run on Windows native only via WSL2

dougsland replied:
humm let's be safe no?

ericcurtin replied:
But if we want to start windows native, I'd suggest we test enable core functionality first, rather than bells and whistles. I'm trying to be mindful of maintenance

ericcurtin replied:
We'd need native Windows CI for one, I don't want to add a very minor part of Windows native enablement and give people false expectations of compatibility.

ericcurtin replied:
We'd need native Windows CI for one, I don't want to add a minor part of Windows native enablement and give people false expectations of compatibility.

ericcurtin replied:
We'd need native Windows CI for one, I don't want to add a minor part of Windows native enablement and give people false expectations of compatibility.

ericcurtin commented on ramalama/console.py:
```diff
@@ -0,0 +1,53 @@
+
+def is_locale_utf8():
+    """Check if the system locale is UTF-8."""
+    lang = os.getenv("LC_CTYPE", "") or os.getenv("LANG", "")
```
Lets just copy systemd 100%, no need to reinvent the wheel.

If the systemd technique doesn't work for whatever reason, we can open PRs against RamaLama and systemd to fix.

ericcurtin replied:
So this means just check for "UTF-8" string

ericcurtin replied:
And we should check LC_CTYPE and LANG separately.

Because of the first or, we won't check both properly.

dougsland replied:
alright

ericcurtin commented on ramalama/console.py:
```diff
@@ -0,0 +1,53 @@
+    return True  # Most Linux/macOS terminals support emoji
+
+# Allow users to override emoji support via an environment variable
+EMOJI = os.getenv("RAMALAMA_FORCE_EMOJI", "1") != "0" and supports_emoji()
```
Are you sure this shouldn't be an or?

FORCE won't force here, because of the AND condition.

Also, surely we should default to 0 and not force by default.

ericcurtin replied:
Basically we want

if (RAMALAMA_FORCE_EMOJI is set) or (supports_emoji())

dougsland commented:
```
export RAMALAMA_FORCE_EMOJI=false
ramalama run deepseek-r1:8b
 38% |█████████████████████████████████████████████████████████████                                                                                                      |    1.79 GB/   4.58 GB   6.33 MB/s   12m 14s2025-02-02 12:30:22,832 - WARNING - [WARNING] TimeoutError: The server took too long to respond. Retrying 1/5...
2025-02-02 12:30:23,036 - WARNING - [WARNING] I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 2/5...
2025-02-02 12:30:23,439 - WARNING - [WARNING] I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 3/5...
2025-02-02 12:30:24,245 - WARNING - [WARNING] I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 4/5...
2025-02-02 12:30:25,848 - WARNING - [WARNING] I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 5/5...
2025-02-02 12:30:25,848 - ERROR - [ERROR] 
Download failed after multiple attempts.
Possible causes:
- Internet connection issue
- Server is down or unresponsive
- Firewall or proxy blocking the request
```

dougsland commented:
> macOS:
> 
> ```
> export RAMALAMA_FORCE_EMOJI=false
> ramalama run deepseek-r1:8b
>  38% |█████████████████████████████████████████████████████████████                                                                                                      |    1.79 GB/   4.58 GB   6.33 MB/s   12m 14s2025-02-02 12:30:22,832 - WARNING - [WARNING] TimeoutError: The server took too long to respond. Retrying 1/5...
> 2025-02-02 12:30:23,036 - WARNING - [WARNING] I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 2/5...
> 2025-02-02 12:30:23,439 - WARNING - [WARNING] I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 3/5...
> 2025-02-02 12:30:24,245 - WARNING - [WARNING] I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 4/5...
> 2025-02-02 12:30:25,848 - WARNING - [WARNING] I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 5/5...
> 2025-02-02 12:30:25,848 - ERROR - [ERROR] 
> Download failed after multiple attempts.
> Possible causes:
> - Internet connection issue
> - Server is down or unresponsive
> - Firewall or proxy blocking the request
> ```
> 
> ```
> $ export RAMALAMA_FORCE_EMOJI=""
> ramalama run deepseek-r1:8b   
>  39% |███████████████████████████████████████████████████████████████                                                                                                    |    1.81 GB/   4.58 GB   9.28 MB/s    8m 23s2025-02-02 12:32:19,210 - WARNING - ⚠️  TimeoutError: The server took too long to respond. Retrying 1/5...
> 2025-02-02 12:32:19,417 - WARNING - ⚠️  I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 2/5...
> 2025-02-02 12:32:19,825 - WARNING - ⚠️  I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 3/5...
> 2025-02-02 12:32:20,633 - WARNING - ⚠️  I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 4/5...
> 2025-02-02 12:32:22,240 - WARNING - ⚠️  I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 5/5...
> 2025-02-02 12:32:22,240 - ERROR - ❌ 
> Download failed after multiple attempts.
> Possible causes:
> - Internet connection issue
> - Server is down or unresponsive
> - Firewall or proxy blocking the request
> ```
> 
> ```
>  $ export RAMALAMA_FORCE_EMOJI="true"
>  $ ramalama run deepseek-r1:8b       
>  39% |███████████████████████████████████████████████████████████████                                                                                                    |    1.82 GB/   4.58 GB   3.25 MB/s   24m 00s2025-02-02 12:33:41,977 - WARNING - ⚠️  TimeoutError: The server took too long to respond. Retrying 1/5...
> 2025-02-02 12:33:42,186 - WARNING - ⚠️  I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 2/5...
> 2025-02-02 12:33:42,592 - WARNING - ⚠️  I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 3/5...
> 2025-02-02 12:33:43,398 - WARNING - ⚠️  I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 4/5...
> 2025-02-02 12:33:45,005 - WARNING - ⚠️  I/O Error: Network error: [Errno 8] nodename nor servname provided, or not known. Retrying 5/5...
> 2025-02-02 12:33:45,005 - ERROR - ❌ 
> Download failed after multiple attempts.
> Possible causes:
> - Internet connection issue
> - Server is down or unresponsive
> - Firewall or proxy blocking the request
> ```

same behaviour on Fedora 41

dougsland commented:
@ericcurtin ready for review, i just didn't follow why I need to change install.sh. Make install seems to work just nice.

ericcurtin reviewed (changes requested):
We need some changes here, for one we cannot break install.sh right now, people are using that to install right now

ericcurtin commented:
"make install" requires a "git clone", that's not a reasonable thing to expect a user to do

ericcurtin commented:
Bug and bug fix reported from lack of updating the install.sh file:

https://github.com/containers/ramalama/pull/704

rhatdan commented on install.sh:
```diff
@@ -89,8 +89,8 @@ setup_ramalama() {
-  local branch="${BRANCH:-s}"
-  local url="${host}/containers/ramalama/${branch}/bin/${from_file}"
+  local branch="main"
+  local url="${host}/kush-gupt/ramalama/${branch}/bin/${from_file}"
```
This looks broken.

ericcurtin commented:
This is all we need to do for emoji-support detection:

```
def is_locale_utf8():
    # List of environment variables to check
    env_vars = ['LC_ALL', 'LC_CTYPE', 'LANG']
    for var in env_vars:
        value = os.environ.get(var, '')
        if 'UTF-8' in value:
            return True

    return False

def emoji_enabled():
    term = os.getenv("TERM")
    if term is None or term in ["dumb", "linux"]:
        return False

    return is_locale_utf8()
```

I didn't invent this technique or anything, I just ported this from systemd, it's more tried and tested than anything any of us would handcraft.

ericcurtin commented:
But the FORCE or emoji_enabled() code with the cached variable looks fine.

ericcurtin commented on ramalama/console.py:
```diff
@@ -0,0 +1,35 @@
+
+def is_locale_utf8():
+    """Check if the system locale is UTF-8."""
+    return 'UTF-8' in os.getenv('LC_CTYPE', '') or 'UTF-8' in os.getenv('LANG', '')
```
This code looks great, LC_ALL is the other or you could add if you wanted

dougsland commented:
> Bug reported and bug fix from lack of updating the install.sh file:
> 
> #704

@ericcurtin install.sh should be improved to avoid breaking things (automatic detect new files or other mechanism). and also added to CI/CD @ericcurtin. Regarding the windows code, everything has a start, we don't need to start big. I just added as mr. @rhatdan requested. If he request to added Windows support, we will make it fly.

dougsland commented:
Let me close this one, I mess with my sync branch request, let me do a fresh PR and link this one.

ericcurtin commented:
@dougsland our Windows solution is not native Windows it's WSL2, I understand Windows support is open to interpretation, I was trying to clarify that, it's documented here:

https://github.com/ollama/ollama/blob/main/docs/windows.md
